src/wazuh_modules/agent_upgrade/wm_agent_upgrade_validate.c

```diff
@@ -0,0 +1,114 @@
+
+/**
+ * Check if agent exist
+ * @param agent_id Id of agent to validate
+ * @return return_code
+ * @retval WM_UPGRADE_SUCCESS_VALIDATE
+ * @retval WM_UPGRADE_NOT_AGENT_IN_DB
+ * */
+int wm_agent_upgrade_validate_id(int agent_id) {
+    char *name = NULL;
+    int return_code = WM_UPGRADE_SUCCESS_VALIDATE;
+    if (agent_id == MANAGER_ID) {
+        return_code = WM_UPGRADE_INVALID_ACTION_FOR_MANAGER;
+    } else if (name = wdb_agent_name(agent_id), name) {
+        // Agent found: OK
+        free(name);
+    } else {
+        return_code = WM_UPGRADE_NOT_AGENT_IN_DB;
+    }
+
+    return return_code;
+}
+
+/**
+ * Check if agent status is active
+ * @param agent_id Id of agent to validate
+ * @return return_code
+ * @retval WM_UPGRADE_SUCCESS_VALIDATE
+ * @retval WM_UPGRADE_AGENT_IS_NOT_ACTIVE
+ * */
+int wm_agent_upgrade_validate_status(int agent_id) {
+    int return_code = WM_UPGRADE_SUCCESS_VALIDATE;
+    int last_keepalive = wdb_agent_last_keepalive(agent_id);
+
+    if (last_keepalive < 0 || last_keepalive < (time(0) - DISCON_TIME)) {
+        return_code = WM_UPGRADE_AGENT_IS_NOT_ACTIVE;
+    }
+
+    return return_code;
+}
+
+/**
+ * Check if agent version is valid to upgrade
+ * @param agent_id Id of agent to validate
+ * @return return_code
+ * @retval WM_UPGRADE_SUCCESS_VALIDATE
+ * @retval WM_UPGRADE_NOT_MINIMAL_VERSION_SUPPORTED
+ * @retval WM_UPGRADE_VERSION_SAME_MANAGER
+ * @retval WM_UPGRADE_NEW_VERSION_LEES_OR_EQUAL_THAT_CURRENT
+ * @retval WM_UPGRADE_NEW_VERSION_GREATER_MASTER)
+ * */
+int wm_agent_upgrade_validate_agent_version(int agent_id, void *task, wm_upgrade_command command) {
+    char *agent_version = NULL;
+    char *tmp_agent_version = NULL;
+    int return_code = WM_UPGRADE_SUCCESS_VALIDATE;
+
+    if (agent_version = wdb_agent_version(agent_id), agent_version) {
+        tmp_agent_version = strchr(agent_version, 'v');
+        
+        if (strcmp(tmp_agent_version, WM_UPGRADE_MINIMAL_VERSION_SUPPORT) < 0) {
+            return_code = WM_UPGRADE_NOT_MINIMAL_VERSION_SUPPORTED;
+        } else if (WM_UPGRADE_UPGRADE == command) {
+            task = (wm_upgrade_task *)task;
+            return_code = wm_agent_upgrade_validate_non_custom_version(tmp_agent_version, task);
+        }
+
+        free(agent_version);
+    }
+
+    return return_code;
+}
+
+/**
+ * Check if agent version is valid to upgrade to a non-customized version
+ * @param agent_id Id of agent to validate
+ * @return return_code
+ * @retval WM_UPGRADE_SUCCESS_VALIDATE
+ * @retval WM_UPGRADE_VERSION_SAME_MANAGER
+ * @retval WM_UPGRADE_NEW_VERSION_LEES_OR_EQUAL_THAT_CURRENT
+ * @retval WM_UPGRADE_NEW_VERSION_GREATER_MASTER)
+ * */
+static int wm_agent_upgrade_validate_non_custom_version(char *agent_version, wm_upgrade_task *task) {
+    char *master_version = NULL;
+    char *tmp_master_version = NULL;
+    master_version = wdb_agent_version(MANAGER_ID);
```

Comment:
I think the name of this variable should be `manager_version`.

Reply:
Also, check if it is not NULL.

Reply:
you're right, solved